File: cpp/p023.cpp
#include <iostream>
#include <vector>

int compute(const int n) {
	std::vector<int> abundant;
	for (int a = 1; a <= n; a++) {
		int a_sum = 0;
		for (int i = 1; i < a; i++) {
			if (a % i == 0)
				a_sum += i;
		}
		if (a_sum > a)
			abundant.push_back(a);
	}
	std::vector<int> sum_abundant(n + 1, 0);
	for (int a = 0; a < abundant.size(); a++) {
		for (int b = a; b < abundant.size(); b++) {
			if (abundant[a] + abundant[b] <= n)
				sum_abundant[abundant[a] + abundant[b]] = 1;
			else
				break;
		}
	}
	int sum = 0;
	for (int i = 1; i < n; i++) {
		if (!sum_abundant[i])
			sum += i;
	}
	return sum;
}

int main() {
	std::cout << compute(28123) << std::endl;
	return 0;
}

File: python/p023.py
def compute(N):
	set = [1] * (N + 1)
	initial_primes = {2, 3, 5, 7, 9, 11, 13}
	for p in initial_primes:
		for i in range(p * p, N + 1, p):
			if set[i] == 1:
				set[i] = 0
	abundant = []
	for a in range(1, N + 1):
		if set[a] == 0: # Don't check prime numbers
			a_sum = 0
			for i in range(1, a):
				if a % i == 0:
					a_sum += i
			if a_sum > a:
				abundant.append(a)
	sum_abundant = [0] * (N + 1)
	for a in range(0, len(abundant)):
		for b in range(a, len(abundant)):
			if abundant[a] + abundant[b] <= N:
				sum_abundant[abundant[a] + abundant[b]] = 1
			else:
				break
	sum = 0
	for i in range(1, N):
		if sum_abundant[i] == 0:
			sum += i
	return sum

if __name__ == '__main__':
	print(compute(28123))
